redis/lookup.go

```diff
@@ -0,0 +1,154 @@
+package redis
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"strings"
+	"time"
+
+	"github.com/letsencrypt/boulder/cmd"
+	blog "github.com/letsencrypt/boulder/log"
+
+	"github.com/redis/go-redis/v9"
+)
+
+// Lookup is a helper that keeps *redis.Ring shards up to date using SRV
+// lookups.
+type Lookup struct {
+	// service is the symbolic name of the desired service.
+	service string
+
+	// domain is the domain name of the desired service.
+	domain string
+
+	// updateFrequency is the frequency of periodic SRV lookups. Defaults to 30
+	// seconds.
+	updateFrequency time.Duration
+
+	// dnsAuthority is the single <hostname|IPv4|[IPv6]>:<port> of the DNS
+	// server to be used for SRV lookups. If the address contains a hostname it
+	// will be resolved via the system DNS. If the port is left unspecified it
+	// will default to '53'. If this field is left unspecified the system DNS
+	// will be used for resolution.
+	dnsAuthority string
+
+	ring   *redis.Ring
+	logger blog.Logger
+}
+
+// NewLookup returns a new Lookup helper.
+func NewLookup(srv cmd.ServiceDomain, dnsAuthority string, frequency time.Duration, ring *redis.Ring, logger blog.Logger) *Lookup {
+	if frequency == 0 {
+		// Use default frequency.
+		frequency = 30 * time.Second
+	}
+	if dnsAuthority != "" {
+		host, port, err := net.SplitHostPort(dnsAuthority)
+		if err != nil {
+			// Assume only hostname or IPv4 address was specified.
+			host = dnsAuthority
+			port = "53"
+		}
+		dnsAuthority = net.JoinHostPort(host, port)
+	}
+	return &Lookup{
+		service:         srv.Service,
+		domain:          srv.Domain,
+		updateFrequency: frequency,
+		dnsAuthority:    dnsAuthority,
+		ring:            ring,
+		logger:          logger,
+	}
+}
+
+// getResolver returns a resolver that will be used to perform SRV lookups.
+func (look *Lookup) getResolver() *net.Resolver {
+	if look.dnsAuthority == "" {
+		return net.DefaultResolver
+	}
+	return &net.Resolver{
+		PreferGo: true,
+		Dial: func(ctx context.Context, network, address string) (net.Conn, error) {
+			return net.Dial(network, look.dnsAuthority)
+		},
+	}
+}
+
+// dnsName returns DNS name to look up as defined in RFC 2782.
+func (look *Lookup) dnsName() string {
+	return fmt.Sprintf("_%s._tcp.%s", look.service, look.domain)
```

Comment:
I think the Lookup struct should contain the protocol too: `tcp` or `udp`. Then this string format and the `Shards` method could use the protocol stored in the struct. RFC 2782 supports both for what it's worth. Currently I can only find SRV records containing `_tcp` in staging and prod though.

Reply:
I'm not sure I see the value here, the Redis protocol is `tcp`-only.

Reply:
🤦🏼 Yup, RESP is TCP only. https://redis.io/docs/reference/protocol-spec/